libs/depgraph/README.md

```diff
@@ -0,0 +1,284 @@
+in the managed system. A graph-wide unique item identifier is therefore a combination
+of the type (returned by `Type() string`) and the name. Every distinct item is modeled
+by the graph as a separate node (`Node` struct).
+Another notable method defined by the `Item` interface is `Dependencies()`. It lists
+all the dependencies of an item and therefore determines the outgoing edges of the
+node representing the item.
+
+Here is a simplified example for Items representing Linux interfaces and routes.
+Only `Name()`, `Type()` and `Dependencies()` methods are shown here. For other
+required methods, please see the `Item` interface definition.
+
+```go
+import "github.com/lf-edge/eve/libs/depgraph"
+
+type LinuxInterface struct {
+    name string
+}
+
+func (intf LinuxInterface) Name() string {
+    return intf.name
+}
+
+func (intf LinuxInterface) Type() string {
+    return "Linux interface"
+}
+
+func (intf LinuxInterface) Dependencies() []depgraph.Dependency {
+    // no dependencies
+    return []depgraph.Dependency{}
+}
+
+// Other Item methods for LinuxInterface are omitted.
+
+type LinuxRoute struct {
+    via    string
+    dst    string
+    metric int
+}
+
+func (route LinuxRoute) Name() string {
+    // Both dst and via should be included to uniquely identify the route
+    // among all the routes.
+    return route.dst + " via " + route.via
+}
+
+func (route LinuxRoute) Type() string {
+    return "Linux route"
+}
+
+func (route LinuxRoute) Dependencies() []depgraph.Dependency {
+    return []depgraph.Dependency{
+        {
+            Item: depgraph.RequiredItem{
+                Type: "Linux interface", // can also use LinuxInterface{}.Type()
+                Name: route.via, // can also use LinuxInterface{name: route.via}.Name()
+            },
+            Description: "Route requires outgoing interface to be configured first",
+        },
+    }
+}
+
+// Other Item methods for LinuxRoute are omitted.
+```
+
+Once all needed `Item` implementations are available, you are ready to build a dependency
+graph to model some kind of system state.
+Graph with some initial content is created using `New()`:
+
+```go
+import "github.com/lf-edge/eve/libs/depgraph"
+
+// using LinuxRoute and LinuxInterface defined above
+
+g := depgraph.New(depgraph.InitArgs{
+    Name:        "MyGraph",
+    Description: "This is my example graph",
+    Items:       []Item{
+        LinuxRoute{via: "eth0", dst: "10.10.0.0/16", metric: 10},
+        LinuxRoute{via: "eth0", dst: "192.168.16.0/24", metric: 10},
+        LinuxInterface{name: "eth0"},
+    },
+})
+```
+
+Graph will automatically create one node for each item and will build edges based on
+dependencies of these items. In the example above, there will be edge from both routes
+pointing to the `eth0` interface.
+`InitArgs` can also contain the initial content of subgraphs.
+
+A single Node is manipulated using `Node()`, `PutNode()` and `DelNode()` methods:
```

Comment:
I found this a bit confusing. Isn't a `Node` just something that is an `Item`? Why would I wrap it? I get why the graph might wrap it internally, but why do we ever use it?

Reply:
I have got rid of the Node from the API. Now it is just an `Item`.
Note that I have added `ItemState` to the API for the caller (like reconciler) to be able to store run-time state information alongside items.